File: app/src/main/java/com/onedelay/mymovie/activities/AllReviewActivity.java
package com.onedelay.mymovie.activities;

import android.app.Activity;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;
import android.widget.Toast;

import com.onedelay.mymovie.Constants;
import com.onedelay.mymovie.R;
import com.onedelay.mymovie.adapters.ReviewAdapter;
import com.onedelay.mymovie.api.RequestProvider;
import com.onedelay.mymovie.database.ReviewEntity;
import com.onedelay.mymovie.utils.DividerItemDecorator;
import com.onedelay.mymovie.viewmodels.ReviewListViewModel;

import java.util.List;

public class AllReviewActivity extends AppCompatActivity {
    private TextView score;
    private ReviewAdapter adapter;

    private float rating;

    private ReviewListViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_all_review);

        // 앱바 제목 텍스트 변경
        ActionBar ab = getSupportActionBar();
        if (ab != null) {
            ab.setTitle(getString(R.string.appbar_review_list));
            ab.setDisplayHomeAsUpEnabled(true);
            ab.setHomeButtonEnabled(true);
        }

        findViewById(R.id.btn_write).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (RequestProvider.isNetworkConnected(getBaseContext())) {
                    Intent intent = new Intent(AllReviewActivity.this, WriteReviewActivity.class);
                    intent.putExtra(Constants.KEY_MOVIE_ID, getIntent().getIntExtra(Constants.KEY_MOVIE_ID, 0));
                    intent.putExtra(Constants.KEY_GRADE, getIntent().getIntExtra(Constants.KEY_GRADE, 12));
                    intent.putExtra(Constants.KEY_TITLE, getIntent().getStringExtra(Constants.KEY_TITLE));
                    startActivityForResult(intent, Constants.WRITE_REQUEST);
                } else {
                    Toast.makeText(AllReviewActivity.this, getResources().getString(R.string.toast_network_error), Toast.LENGTH_SHORT).show();
                }
            }
        });

        RecyclerView recyclerView = findViewById(R.id.review_list);

        adapter = new ReviewAdapter(getApplicationContext());
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
        recyclerView.setHasFixedSize(true);

        // 구분선
        RecyclerView.ItemDecoration dividerItemDecoration = new DividerItemDecorator(ContextCompat.getDrawable(getApplicationContext(), R.drawable.recyclerview_divider));
        recyclerView.addItemDecoration(dividerItemDecoration);

        viewModel = ViewModelProviders.of(this).get(ReviewListViewModel.class);
        if (RequestProvider.isNetworkConnected(getBaseContext())) {
            viewModel.requestReviewList(getIntent().getIntExtra(Constants.KEY_MOVIE_ID, 0));
        }

        viewModel.setData(getIntent().getIntExtra(Constants.KEY_MOVIE_ID, 0));
        viewModel.getData().observe(this, new Observer<List<ReviewEntity>>() {
            @Override
            public void onChanged(@Nullable List<ReviewEntity> reviews) {
                if (reviews != null) {
                    score.setText(String.format(getString(R.string.all_review_score), rating, reviews.size()));
                    adapter.setItems(reviews);
                    adapter.notifyDataSetChanged();
                }
            }
        });

        TextView textView = findViewById(R.id.movie_title);
        textView.setText(getIntent().getStringExtra(Constants.KEY_TITLE));

        ImageView imageView = findViewById(R.id.level);
        int grade = getIntent().getIntExtra(Constants.KEY_GRADE, 12);

        switch (grade) {
            case 12:
                imageView.setImageResource(R.drawable.ic_12);
                break;
            case 15:
                imageView.setImageResource(R.drawable.ic_15);
                break;
            case 19:
                imageView.setImageResource(R.drawable.ic_19);
                break;
            default:
                imageView.setImageResource(R.drawable.ic_all);
        }

        rating = getIntent().getFloatExtra(Constants.KEY_RATING, 0.0f);

        RatingBar ratingBar = findViewById(R.id.rating_bar);
        ratingBar.setRating(rating / 2);

        score = findViewById(R.id.score);

        adapter.setOnItemClickListener(new ReviewAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(int position) {
                Toast.makeText(AllReviewActivity.this, getResources().getString(R.string.toast_reporting), Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onRecommendClick(final int position, final int value) {
                if (RequestProvider.isNetworkConnected(getBaseContext())) {
                    viewModel.requestReviewRecommend(getIntent().getIntExtra(Constants.KEY_MOVIE_ID, 0), adapter.getItem(position).getId(), "onedelay", adapter);
                } else {
                    Toast.makeText(getBaseContext(), getResources().getString(R.string.toast_network_error), Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return true;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == Constants.WRITE_REQUEST && resultCode == Activity.RESULT_OK) {
            viewModel.requestReviewList(getIntent().getIntExtra(Constants.KEY_MOVIE_ID, 0));
        }
    }
}

File: app/src/main/java/com/onedelay/mymovie/activities/AllReviewActivity.kt
package com.onedelay.mymovie.activities

import android.app.Activity
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.view.MenuItem
import android.widget.Toast
import com.onedelay.mymovie.Constants
import com.onedelay.mymovie.R
import com.onedelay.mymovie.adapters.ReviewAdapter
import com.onedelay.mymovie.data.api.RequestProvider
import com.onedelay.mymovie.utils.DividerItemDecorator
import com.onedelay.mymovie.viewmodels.ReviewListViewModel
import kotlinx.android.synthetic.main.activity_all_review.*

class AllReviewActivity : AppCompatActivity(), ReviewAdapter.OnItemClickListener {
    private var rating: Float = 0F

    private var viewModel: ReviewListViewModel? = null

    private val adapter = ReviewAdapter(this@AllReviewActivity)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_all_review)
        initViewModel(ViewModelProviders.of(this).get(ReviewListViewModel::class.java))
        initView()
        rating = intent.getFloatExtra(Constants.KEY_RATING, 0.0f)
    }

    private fun initView() {
        supportActionBar?.apply {
            title = getString(R.string.appbar_review_list)
            setDisplayHomeAsUpEnabled(true)
            setHomeButtonEnabled(true)
        }

        btn_write.setOnClickListener {
            if (RequestProvider.isNetworkConnected(baseContext)) {
                val intent = Intent(this@AllReviewActivity, WriteReviewActivity::class.java).apply {
                    putExtra(Constants.KEY_MOVIE_ID, intent.getIntExtra(Constants.KEY_MOVIE_ID, 0))
                    putExtra(Constants.KEY_GRADE, intent.getIntExtra(Constants.KEY_GRADE, 12))
                    putExtra(Constants.KEY_TITLE, intent.getStringExtra(Constants.KEY_TITLE))
                }
                startActivityForResult(intent, Constants.WRITE_REQUEST)
            } else {
                Toast.makeText(this@AllReviewActivity, resources.getString(R.string.toast_network_error), Toast.LENGTH_SHORT).show()
            }
        }

        review_list.apply {
            adapter = this@AllReviewActivity.adapter
            layoutManager = LinearLayoutManager(baseContext)
            setHasFixedSize(true)

            ContextCompat.getDrawable(baseContext, R.drawable.recyclerview_divider)?.let {
                DividerItemDecorator(it)
            }.also { addItemDecoration(it) }
        }

        movie_title.text = intent.getStringExtra(Constants.KEY_TITLE)

        val grade = intent.getIntExtra(Constants.KEY_GRADE, 12)
        when (grade) {
            12 -> level.setImageResource(R.drawable.ic_12)
            15 -> level.setImageResource(R.drawable.ic_15)
            19 -> level.setImageResource(R.drawable.ic_19)
            else -> level.setImageResource(R.drawable.ic_all)
        }

        rating_bar.rating = rating / 2
    }

    private fun initViewModel(viewModel: ReviewListViewModel) {
        this.viewModel = viewModel
        if (RequestProvider.isNetworkConnected(baseContext)) {
            viewModel.requestReviewList(intent.getIntExtra(Constants.KEY_MOVIE_ID, 0))
        }

        val movieId = intent.getIntExtra(Constants.KEY_MOVIE_ID, 0)
        viewModel.setData(intent.getIntExtra(Constants.KEY_MOVIE_ID, 0))
        viewModel.getReviewList(movieId).observe(this, Observer { reviews ->
            if (reviews != null) {
                score.text = String.format(getString(R.string.all_review_score), rating, reviews.size)
                adapter.setItems(reviews)
            }
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
        }
        return true
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == Constants.WRITE_REQUEST && resultCode == Activity.RESULT_OK) {
            viewModel?.requestReviewList(intent.getIntExtra(Constants.KEY_MOVIE_ID, 0))
        }
    }

    override fun onItemClick(position: Int) {
        Toast.makeText(this@AllReviewActivity, resources.getString(R.string.toast_reporting), Toast.LENGTH_SHORT).show()
    }

    override fun onRecommendClick(position: Int, value: Int) {
        if (RequestProvider.isNetworkConnected(baseContext)) {
            viewModel?.requestReviewRecommend(intent.getIntExtra(Constants.KEY_MOVIE_ID, 0), adapter?.getItem(position).id, "onedelay", adapter)
        } else {
            Toast.makeText(baseContext, resources.getString(R.string.toast_network_error), Toast.LENGTH_SHORT).show()
        }
    }
}
